src/lib/define-smart-component.js

```diff
@@ -0,0 +1,98 @@
+import { produce } from './immer.js';
+import { defineSmartComponentCore } from './smart-manager.js';
+
+const META = Symbol('META');
+
+/**
+ * @typedef {Element} SmartContainer
+ * @param {object} context
+ */
+
+/**
+ * @param {object} definition
+ * @param {string} definition.selector
+ * @param {object} definition.params
+ * @param {({
```

Comment:
When using:

```javascript
defineSmartComponent({
  onContextUpdate ({ updateComponent }) {
  },
  selector: 'selector',
  params: {},
}, null);
```

WebStorm is complaining because the `onContextUpdate` is not destructuring all the properties.
Making the properties optional in JSDoc makes WebStorm happy:

```javascript
 * @param {({
 *   container?: SmartContainer,
 *   component?: Element,
 *   context?: object,
 *   onEvent?: (type: string, listener: (detail: object) => void) => void,
 *   updateComponent?: (
 *     type: string,
 *     callbackOrObject: ((property: object) => void),
 *   ) => void,
 *   signal?: AbortSignal,
 * }) => void} definition.onContextUpdate

```

Reply:
I applied this but we may mark some of them as required in the future (once we have better typechecking).